at_lookup/lib/src/cache/cacheable_secondary_address_finder.dart

```diff
@@ -0,0 +1,194 @@
+          secondaryUrl == 'data:null') {
+        throw SecondaryNotFoundException(getNotFoundExceptionMessage(atSign));
+      }
+      var secondaryInfo = LookUpUtil.getSecondaryInfo(secondaryUrl);
+      String secondaryHost = secondaryInfo[0];
+      int secondaryPort = int.parse(secondaryInfo[1]);
+
+      SecondaryAddress addr = SecondaryAddress(secondaryHost, secondaryPort);
+      _map[atSign] = SecondaryAddressCacheEntry(
+          addr, DateTime.now().add(cacheFor).millisecondsSinceEpoch);
+    } on Exception catch (e) {
+      _logger
+          .severe('${getNotFoundExceptionMessage(atSign)} - ${e.toString()}');
+      rethrow;
+    }
+  }
+}
+
+class SecondaryAddressCacheEntry {
+  final SecondaryAddress secondaryAddress;
+
+  /// milliseconds since epoch
+  final int expiresAt;
+
+  SecondaryAddressCacheEntry(this.secondaryAddress, this.expiresAt);
+}
+
+class SecondaryUrlFinder {
+  final String _rootDomain;
+  final int _rootPort;
+  SecondaryUrlFinder(this._rootDomain, this._rootPort);
+
+  Future<String?> findSecondaryUrl(String atSign) async {
+    return await _findSecondary(atSign);
+  }
+
+  Future<String?> _findSecondary(String atsign) async {
+    String? response;
+    SecureSocket? socket;
+    try {
+      AtSignLogger('AtLookup')
+          .finer('AtLookup.findSecondary received atsign: $atsign');
+      if (atsign.startsWith('@')) atsign = atsign.replaceFirst('@', '');
+      var answer = '';
+      String? secondary;
+      var ans = false;
+      var prompt = false;
+      var once = true;
+      // ignore: omit_local_variable_types
+      socket = await SecureSocket.connect(_rootDomain, _rootPort);
+      // listen to the received data event stream
+      socket.listen((List<int> event) async {
+        answer = ''; //TODO utf8.decode(event);
+
+        if (answer.endsWith('@') && prompt == false && once == true) {
+          prompt = true;
+          socket!.write('$atsign\n');
+          await socket.flush();
+          once = false;
+        }
+
+        if (answer.contains(':')) {
+          answer = answer.replaceFirst('\r\n@', '');
+          answer = answer.replaceFirst('@', '');
+          answer = answer.replaceAll('@', '');
+          secondary = answer.trim();
+          ans = true;
+        } else if (answer.startsWith('null')) {
+          secondary = null;
+          ans = true;
+        }
+      });
+      // wait 30 seconds
+      for (var i = 0; i < 6000; i++) {
+        await Future.delayed(Duration(milliseconds: 5));
+        if (ans) {
+          response = secondary;
+          socket.write('@exit\n');
+          await socket.flush();
+          socket.destroy();
+          AtSignLogger('AtLookup').finer(
+              'AtLookup.findSecondary got answer: $secondary and closing connection');
+          return response;
+        }
+      }
+      // .. and close the socket
+      await socket.flush();
+      socket.destroy();
+      throw Exception('AtLookup.findSecondary timed out');
```

Comment:
According to our exception hierarchy, I recon this corresponds to TimeoutException. We will revisit this in a different PR.

Reply:
made the change to throw exact exceptions